### PR TITLE
Authentication

### LINKED ISSUE #14: Authentication
Create USER authentication middleware.

Use OAuth for authentication.

**Make sure any sensitive information is securely stored!**

> Rubric goal: OAuth or another approved form of third party authentication is implemented and is customized or embedded such that it appears to be a fully integrated part of the application.

### PR DESCRIPTION
**Pull request by Pav0l**
# Description

Please include a summary of the change and a link to which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes #14  #12 

Created User authentication with JWT.
Created auth routes to create new user and login user.

**Installed dependencies:**
- jswonwebtoken
- bcrypt

**TO DO:**

- [ ] Write automatic tests
- [ ] Create documentation for endpoints
- [x] Resolve any merge conflicts

## Type of change

Please add a relevant label.

## Change status

- [X] Complete, but not tested (may need new tests)

# How Has This Been Tested?

Please describe steps taken to ensure this feature is functional and does not break other features:

Features were tested manually with Postman and local MongoDB setup with Compass. All endpoints return proper values or error messages. Local DB is updated with proper data.

# Checklist:

- [x] My code follows the style guidelines of this project

- [x] I have performed a self-review of my own code

- [x] I have commented my code, particularly in hard-to-understand areas

- [ ] I have made corresponding changes to the documentation, if relevant

- [x] I have run the app using my feature and ensured that no functionality is broken

- [x] My changes generate no new warnings

- [x] There are no merge conflicts

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,17 +1,28 @@
...
-const { connectDb } = require('./models/index');
+const { models, connectDb } = require('./models/index');
 require('dotenv').config();
 
 const port = process.env.PORT || 7000;
 
 connectDb()
   .then(async () => {
     try {
-      if (process.env.NODE_ENV === 'development' && usersLength.length === 0) {
```
Comment (DianaIgnatescu) [inline index.js:12]: Can you please comment this out?

File: routes/auth.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+  '/register',
+  // add validation middleware
+  async (req, res, next) => {
+    try {
+      // Check if this name already exist in DB
+      let { name, hotel_id, password } = req.body;
+      const [existingUser] = await models.User.where({ name });
+
+      if (existingUser) {
+        return res.status(400).json({
```
Comment (Pav0l) [inline routes/auth.js]: 422 vs 400 status code

File: index.js (hunk 1 above)
Comment (Pav0l) [reply index.js:12]: Done on the next line

**Review by DianaIgnatescu — APPROVED** [review]
Awesome work, Pavol!

File: middleware/auth.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+const jwtKey = process.env.JWT_SECRET || super_secret;
+
+module.exports = (req, res, next) => {
+  const token = req.headers.authorization;
+
+  if (token) {
+    jwt.verify(token, jwtKey, (err, decoded) => {
+      if (err) {
+        console.error(err);
+        res.status(401).json({ message: 'Invalid Credentials' });
```
Comment (mark-marshall) [inline middleware/auth.js]: Can we add these errors to the util and remove hardcoding in here?

File: utils/secrets.js [hunk 4]
```diff
@@ -0,0 +1,3 @@
+module.exports = {
+  super_secret: 'There_Is_N0_S3cr3t_So_Stop_Looking',
```
Comment (mark-marshall) [inline utils/secrets.js:2]: amazing.

File: utils/createToken.js [hunk 5]
```diff
@@ -0,0 +1,12 @@
+const jwt = require('jsonwebtoken');
+const { super_secret } = require('../utils/secrets');
+
+const createToken = ({ user_id, name, hotel_id, chat_id }) =>
```
Comment (mark-marshall) [inline utils/createToken.js]: Might want to remove chat_id here as it'l be undefined for staff.

File: routes/auth.js [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        });
+      }
+
+      password = bcrypt.hashSync(password, 10);
+
+      // add new user to the DB, rewrite the password to be the hashed pw
+      const user = await models.User.create({ ...req.body, password });
+
+      // remove password from the returned user object, so it's not sent to FE
+      user.password = undefined;
```
Comment (mark-marshall) [inline routes/auth.js:28]: How big is the user object, could you alternatively send a new Obj and just add all other keys except pwrd: { id, name, hotel_id}

File: routes/auth.js [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+      // generate token
+      const token = createToken({
+        id: user.id,
+        name,
+        hotel_id,
+        chat_id: user.chat_id,
```
Comment (mark-marshall) [inline routes/auth.js]: Same with chat_id again here as mentioned previously (and for others going forward if you decide to remove)

File: routes/auth.js [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+const routes = express.Router();
+
+routes.post(
+  '/register',
+  // add validation middleware
+  async (req, res, next) => {
+    try {
+      // Check if this name already exist in DB
+      let { name, hotel_id, password } = req.body;
```
Comment (mark-marshall) [inline routes/auth.js]: the documentExists helper would make this endpoint a fraction more concise but barely worth it - you decide

File: routes/auth.js [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+      // send the user info and token in response
+      res.status(201).json({ user, token });
+    } catch (err) {
+      next(err);
+    }
+  }
+);
+
+routes.post(
```
Comment (mark-marshall) [inline routes/auth.js:45]: If we have different endpoints for staff and guest auth, should we reserve /login for guests, it being the cleanest one?

**Review by mark-marshall — CHANGES_REQUESTED** [review]

File: utils/createToken.js (hunk 5 above)
Comment (Pav0l) [reply utils/createToken.js]: Done 👍 , thanks!

File: routes/auth.js (hunk 9 above)
Comment (Pav0l) [reply routes/auth.js:45]: The path, which user sees will be set on fronted with React Router.

File: middleware/auth.js (hunk 3 above)
Comment (Pav0l) [reply middleware/auth.js]: Will do, after your PR is merged, so I don't cause any merge conflicts.

File: routes/auth.js (hunk 9 above)
Comment (mark-marshall) [reply routes/auth.js:45]: brb - finding way to delete hideous PR comment

**Review by mark-marshall — APPROVED** [review]

**Pull request merged**